fl4health/checkpointing/server_module.py

```diff
@@ -0,0 +1,580 @@
+                return self.state_checkpointer.load_checkpoint(state_checkpoint_name)
+            else:
+                log(INFO, "State checkpointer is defined but no state checkpoint exists.")
+                return None
+        else:
+            raise ValueError("Attempting to load state, but no state checkpointer is specified")
+
+
+class PackingServerCheckpointAndAndStateModule(BaseServerCheckpointAndStateModule):
+    def __init__(
+        self,
+        model: nn.Module | None = None,
+        parameter_exchanger: FullParameterExchangerWithPacking | None = None,
+        model_checkpointers: CheckpointModuleInput = None,
+        state_checkpointer: PerRoundStateCheckpointer | None = None,
+    ) -> None:
+        """
+
+        This module is meant to be a base class for any server-side checkpointing module that relies on unpacking
+        of parameters to hydrate models for checkpointing. The specifics of the unpacking will be handled by the
+        child classes of the packer within the parameter exchange.
+        NOTE: This function ASSUMES full parameter exchange unpacking. If more complex unpacking/parameter exchange
+        is used, this is not the right parent class.
+
+        Args:
+            model (nn.Module | None, optional): Model architecture to be saved. The module will use this architecture
+                to hold the server parameters and facilitate checkpointing with the help of the parameter exchanger.
+                Recall that servers only have parameters rather than torch models. So we need to know where to route
+                these parameters to allow for real models to be saved. Defaults to None.
+            parameter_exchanger (FullParameterExchangerWithPacking | None, optional): This will facilitate routing the
+                server parameters into the right components of the provided model architecture. It specifically also
+                should handle any necessary unpacking of the parameters. Note that this exchanger and the model must
+                match the one used for training and exchange with the servers to ensure parameters go to the right
+                places. Defaults to None.
+            model_checkpointers (CheckpointModuleInput, optional): If defined, this checkpointer (or sequence of
+                checkpointers) is used to checkpoint models based on their defined scoring function. Defaults to None.
+            state_checkpointer (PerRoundStateCheckpointer | None, optional): If defined, this checkpointer will be
+                used to preserve FL training state to facilitate restarting training if interrupted. Generally, this
+                checkpointer will save much more than just the model being trained. Defaults to None.
+        """
+        if parameter_exchanger is not None:
+            assert isinstance(
+                parameter_exchanger, FullParameterExchangerWithPacking
```

Comment:
Why the need for all the structure and classes with the parameter and packing exchangers? From what I can tell we're trying to allow users to pass additional information when pushing and pulling the parameters. They define a packer, which is passed to the packing exchanger subclass which now is also passed here. There just seems to be a lot of classes involved. Can we just have all the packer classes be subclasses of the parameter exchanger? Is there a way to reduce the number of classes and subclasses across the state checkpointers and module checkpointers?